### PR TITLE
Workflow step definitions: action validation

### PR DESCRIPTION
**Pull request by KevinVdV**
This PR is a continuation of: https://github.com/DSpace/DSpace/pull/2646

When a workflow action is performed which is not authorized in the current step, deny performing this action, this ensures:
- Reject an item should not be possible if the current workflow action doesn't have the submit_reject option.
- Edit metadata should not be possible if the current workflow action doesn't have the submit_edit_metadata option.
- Uploading of a file should not be possible if the current workflow action doesn't have the submit_edit_metadata option.
- ...

### CODE REVIEW

**Comment by tdonohue** [comment]
@KevinVdV or @MarieVerdonck : Now that #2646 is merged, this PR can be rebased to make it smaller & easier to review

**Comment by KevinVdV** [comment]
@tdonohue The latest master has been merged into this PR. Ready for your review.

**Review by benbosman — APPROVED** [review]
I've tested this functionality, and was able to confirm an item in the finaleditstep (which doesn't offer a Reject option) now returns a status 422 instead of 204 when trying to reject. Approving the item still works. Rejecting an item in the editstep (which does offer a Reject option) does work as well.I've tested this functionality, and was able to confirm an item in the finaleditstep (which doesn't offer a Reject option) now returns a status 422 instead of 204 when trying to reject. Approving the item still works. Rejecting an item in the editstep (which does offer a Reject option) does work as well.

I would suggest 2 changes to TaskRestRepositoriesIT (but nothing problematic):
* There are a lot of whitespace changes, which is in general not recommended
* The tests of the second and third workflow step would be shorted, faster and easier to read if you'd use a collection with only a workflow group for the second or third workflow step

It would be best if they can be solved in this PR, but since they're limited to layout and speed of tests, I'm already approving this PR

**Comment by MarieVerdonck** [comment]
@benbosman 
TaskRestRepositoriesIT
- Indentation whitespace changes reimported from master
- Tests for 2nd/3rd step limited to just those steps

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/WorkflowItemRestRepository.java [hunk 1]
```diff
@@ -294,4 +317,30 @@ protected void delete(Context context, Integer id) {
...
+            }
+            Workflow workflow = workflowFactory.getWorkflow(claimedTask.getWorkflowItem().getCollection());
+            Step step = workflow.getStep(claimedTask.getStepID());
+            WorkflowActionConfig currentActionConfig = step.getActionConfig(claimedTask.getActionID());
+            if (!currentActionConfig.getProcessingAction().getOptions().contains(SUBMIT_EDIT_METADATA)) {
+                throw new UnprocessableEntityException(SUBMIT_EDIT_METADATA + " is not a valid option on this " +
+                    "action (" + currentActionConfig.getProcessingAction().getClass() + ").");
+            }
+        } catch (SQLException | WorkflowConfigurationException e) {
+            throw new RuntimeException(e.getMessage(), e);
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/WorkflowItemRestRepository.java]: This should provide a more useful error message or just be updated to `new RuntimeException(e);`  Simply returning `e.getMessage()` isn't useful here.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java [hunk 2]
```diff
@@ -1439,7 +1449,7 @@ public void approvalForbiddenTest() throws Exception {
                                            .withName("Sub Community")
                                            .build();
 
-        // the reviewer2 is a reviewer in a different step for the col1 and with the same role than reviewer1 for
+        // the reviewer2 is a reviewer in a different step for the colAA1 and with the same role than reviewer1 for
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java]: Not sure why this comment needs changing? It looked correct, so I think this change can be reverted.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java [hunk 3]
```diff
@@ -1613,7 +1623,7 @@ public void rejectForbiddenTest() throws Exception {
...
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java]: Same here, this comment change can be reverted.

**Review by tdonohue — COMMENTED** [review]
@KevinVdV and @MarieVerdonck : Overall this looks good to me.  Thank you very much for the thorough ITs, and the inline comments (which were very helpful in verifying what the ITs were attempting to achieve).

I have a few minor, inline comments.  Beyond that, I think this looks good to merge.

File: dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/Action.java [hunk 4]
```diff
@@ -50,6 +50,20 @@ public abstract ActionResult execute(Context c, XmlWorkflowItem wfi, Step step,
      */
     public abstract List<String> getOptions();
 
+    /**
+     * Returns true if one of the options is a parameter of the request
+     * @param request   Action request
+     * @return  true if one of the options is a parameter of the request; false if none was found
+     */
+    protected boolean isInOptions(HttpServletRequest request) {
```
[the remark this answers is not in the document]
Comment (MarieVerdonck) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/Action.java]: Renamed

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/WorkflowItemRestRepository.java (hunk 1 above)
Comment (MarieVerdonck) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/WorkflowItemRestRepository.java]: All generic exceptions in this repo given relevant messages

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java (hunk 2 above)
Comment (MarieVerdonck) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java]: From restoring from (for the indentations), re-implemented correction

**Comment by MarieVerdonck** [comment]
@tdonohue All suggestion implemented in latest commit

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now. Thanks @MarieVerdonck for the quick fixes.

**Pull request merged**